### PR TITLE
Proposals v2. Iteration3

### PR DESCRIPTION
**Pull request by shamil-gadelshin**
### Iteration 3 implemented features and changes

- grace period
- threshold - a minimum percentage of quorum that must vote to pass the proposal
- clear redundant substrate dependencies
- add maximum allowed active proposals 
- runtime upgrade proposal type
- add error enum in a codex
- update finalized_at proposal field
- change ProposalStatus type
- add stake
- add slashes (status, quorum, threshold). [Issue will be fixed: "Fix stake module and allow slashing and unstaking in the same block."](https://github.com/Joystream/substrate-runtime-joystream/issues/161)
- cancelation fee & rejection fee for the proposer
- stake mocks

### CODE REVIEW

File: modules/proposals/engine/src/lib.rs [hunk 1]
```diff
@@ -274,91 +316,233 @@ impl<T: Trait> Module<T> {
         <system::Module<T>>::block_number()
     }
 
+    // Enumerates through active proposals. Tally Voting results.
+    // Returns proposals with finalized status and id
+    fn get_finalized_proposals() -> Vec<FinalizedProposal<T>> {
```
Comment (bedeho) [inline modules/proposals/engine/src/lib.rs]: This routine, and the associated types, was really hard to grapple with. Some of it is undoubtedly due to some names that are suboptimal, such as `FinalizedProposal` not actually being a proposal, but an alias for `FinalizedProposalData`, despite the name `get_finalized_proposals`. Also `Proposal::define_proposal_decision_status` does not give a `ProposalDecisionStatus`, but an option of this. On top of this, there are private types like `ProposalStatusResolution`, which needs both a `Proposal`and specific fields in the proposal, like `approvals`, `slashes` etc.

In general, this is all to say, I still can't fully work out if this is entirely correct and efficient, but I suspect it is, do you see any way to simplify and clarify this whole section?

**Review by bedeho — CHANGES_REQUESTED** [review]
- Is it the case that this follows the approach of propagating errors due to broken invariants? e.g. in

```Rust
        let stake_id = stake_balance
            .map(|stake_amount| {
                T::StakeHandlerProvider::stakes().create_stake(stake_amount, account_id)
            })
            .transpose()?;
```

Basically the approach of absorbing broken invariants into the errors? If so, then I think its OK for now, but we should review and conclude on this issue based on the summary from all we learned from the Parity discussion, and have a clear rule to follow.

As is, it is statically possible to fail past the marked _safe mutation_ in the code path, is it guaranteed that the only failures we get from calls to staking library are due to broken invariants, or are some still possible without this, in which case there is a bug in the extrinsics checks in this module.

Another issue with the error propagation approach here is that, since the errors are not properly typed, there is no explicit conversion of error types, and so understanding the full up to date view possible errors requires chasing down error strings in the staking module.

- None of the proposal creation modules guard the origin, I think we want something like `EnsureMemershipOrigin`, which returns the member id. Then this id can be passed into the proposal engine, so that it can use this as the proposer id it compares to the result of its `EnsureProposerOrigin`. But perhaps I am going to far ahead here.

File: modules/proposals/engine/src/types/mod.rs [hunk 2]
```diff
@@ -0,0 +1,720 @@
...
+    pub title: Vec<u8>,
+
+    /// Proposal body
+    pub body: Vec<u8>,
+
+    /// When it was created.
+    pub created_at: BlockNumber,
+
+    /// When it was approved.
+    pub approved_at: Option<BlockNumber>,
```
[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply modules/proposals/engine/src/types/mod.rs]: Removed approved_at.

File: modules/proposals/engine/src/types/mod.rs [hunk 3]
```diff
@@ -0,0 +1,720 @@
...
+
+    /// Current proposal status
+    pub status: ProposalStatus,
+
+    /// Curring voting result for the proposal
+    pub voting_results: VotingResults,
+
+    /// Proposal finalization block number
+    pub finalized_at: Option<BlockNumber>,
```
[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply modules/proposals/engine/src/types/mod.rs]: I moved it to the finalization status. And renamed.

File: modules/proposals/codex/src/lib.rs [hunk 4]
```diff
@@ -20,36 +20,141 @@ mod tests;
...
+            origin,
+            title: Vec<u8>,
+            body: Vec<u8>,
+            text: Vec<u8>,
+            stake_balance: Option<BalanceOf<T>>,
+        ) {
             let parameters = crate::ProposalParameters {
-                voting_period: T::BlockNumber::from(3u32),
-                approval_quorum_percentage: 49,
+                voting_period: T::BlockNumber::from(50000u32),
```
[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply modules/proposals/codex/src/lib.rs]: Moved to the separate module as we agreed.

File: modules/proposals/codex/src/lib.rs [hunk 5]
```diff
@@ -20,36 +20,141 @@ mod tests;
...
+
+        /// Create runtime upgrade proposal type. On approval prints its content.
+        pub fn create_runtime_upgrade_proposal(
+            origin,
+            title: Vec<u8>,
+            body: Vec<u8>,
+            wasm: Vec<u8>,
+            stake_balance: Option<BalanceOf<T>>,
+        ) {
+            let parameters = crate::ProposalParameters {
```
[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply modules/proposals/codex/src/lib.rs]: Moved to the separate module as we agreed.

File: modules/proposals/codex/src/lib.rs [hunk 6]
```diff
@@ -20,36 +20,141 @@ mod tests;
...
+/// Balance alias for staking
+pub type NegativeImbalance<T> =
+    <<T as stake::Trait>::Currency as Currency<<T as system::Trait>::AccountId>>::NegativeImbalance;
+
+// Defines max allowed text proposal text length. Can be override in the config.
+const DEFAULT_TEXT_PROPOSAL_MAX_LEN: u32 = 20_000;
+// Defines max allowed text proposal text length. Can be override in the config.
+const DEFAULT_RUNTIME_PROPOSAL_WASM_MAX_LEN: u32 = 20_000;
+
+decl_error! {
```
[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply modules/proposals/codex/src/lib.rs]: Made a note to move engine and discussion modules to this approach during the next iteration.

[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply modules/proposals/codex/src/lib.rs]: Made a note to move engine and discussion modules to this approach during the next iteration.

File: modules/proposals/codex/src/proposal_types/text_proposal.rs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+use crate::{ProposalExecutable, ProposalType};
+
+/// Text (signal) proposal executable code wrapper. Prints its content on execution.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug, Default)]
+pub struct TextProposalExecutable {
+    /// Text proposal title
+    pub title: Vec<u8>,
+
+    /// Text proposal body (description)
+    pub body: Vec<u8>,
```
Comment (siman) [inline modules/proposals/codex/src/proposal_types/text_proposal.rs]: Names `body` and `text` look confusing to me. Just saying.

File: modules/proposals/engine/src/types/stakes.rs [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    fn slash(&self, stake_id: T::StakeId, slash_balance: BalanceOf<T>) -> Result<(), &'static str>;
+}
+
+/// Default implementation of the stake logic. Uses actual stake module.
+/// 'marker' responsible for the 'Trait' binding.
+pub(crate) struct DefaultStakeHandler<T> {
+    pub marker: PhantomData<T>,
+}
+
+impl<T: Trait> StakeHandler<T> for DefaultStakeHandler<T> {
```
[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply modules/proposals/engine/src/types/stakes.rs]: I added a general comment.

File: modules/proposals/engine/src/types/stakes.rs [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    fn stakes() -> Rc<dyn StakeHandler<T>> {
+        Rc::new(DefaultStakeHandler {
+            marker: PhantomData::<T>::default(),
+        })
+    }
+}
+
+/// Stake logic handler.
+#[cfg_attr(test, automock)] // attributes creates mocks in tesing environment
+pub trait StakeHandler<T: Trait> {
```
[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply modules/proposals/engine/src/types/stakes.rs]: I added a general comment.

**Comment by shamil-gadelshin** [comment]
Refactor stakes

- change StakeHandler trait and its default implementation to simple ‘stake module’ proxy
- introduce ProposalStakeManager to combine calls from the StakeHandler into the simple workflow

File: modules/proposals/codex/src/proposal_types/text_proposal.rs (hunk 7 above)
Comment (shamil-gadelshin) [reply modules/proposals/codex/src/proposal_types/text_proposal.rs]: Will rename to the "description"!

File: modules/proposals/codex/src/lib.rs [hunk 10]
```diff
@@ -20,36 +20,141 @@ mod tests;
...
+/// Balance alias
+pub type BalanceOf<T> =
+    <<T as stake::Trait>::Currency as Currency<<T as system::Trait>::AccountId>>::Balance;
+
+/// Balance alias for staking
+pub type NegativeImbalance<T> =
+    <<T as stake::Trait>::Currency as Currency<<T as system::Trait>::AccountId>>::NegativeImbalance;
+
+// Defines max allowed text proposal text length. Can be override in the config.
+const DEFAULT_TEXT_PROPOSAL_MAX_LEN: u32 = 20_000;
```
[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply modules/proposals/codex/src/lib.rs]: Done

File: modules/proposals/engine/src/types/mod.rs [hunk 11]
```diff
@@ -0,0 +1,720 @@
...
+
+/// Final proposal status and potential error.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub struct FinalizationStatus {
+    /// Final proposal status
+    pub proposal_status: ProposalDecisionStatus,
+
+    /// Error occured during the proposal finalization
+    pub finalization_error: Option<Vec<u8>>,
```
[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply modules/proposals/engine/src/types/mod.rs]: I will fix it in the next iteration after talk with Mokhtar.

File: modules/proposals/engine/src/types/stakes.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply modules/proposals/engine/src/types/stakes.rs]: Refactored.

File: modules/proposals/engine/src/types/stakes.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply modules/proposals/engine/src/types/stakes.rs]: Refactored.

File: modules/proposals/engine/src/types/mod.rs [hunk 12]
```diff
@@ -0,0 +1,720 @@
...
+    pub status: ProposalStatus,
+
+    /// Curring voting result for the proposal
+    pub voting_results: VotingResults,
+
+    /// Proposal finalization block number
+    pub finalized_at: Option<BlockNumber>,
+
+    /// Created stake id for the proposal
+    pub stake_id: Option<StakeId>,
```
[the remark this answers is not in the document]
Comment (bedeho) [reply modules/proposals/engine/src/types/mod.rs]: To be reviwed after discussion with mokhtar.

**Review by bedeho — APPROVED** [review]
Nice work!

**Pull request merged**